zebra-scan/tests/scanner.rs

```diff
@@ -0,0 +1,257 @@
+        "--sapling-keys-to-scan",
+        key,
+        "--listen-addr",
+        listen_addr,
+        "--zebra-rpc-listen-addr",
+        rpc_listen_addr
+    ];
+
+    // Start the scaner using another test directory.
+    let mut zebra_scanner: TestChild<TempDir> = testdir()?.spawn_child(args.clone(), false)?;
+
+    // Check scanner was started.
+    zebra_scanner.expect_stdout_line_matches("loaded Zebra scanner cache")?;
+
+    // The first time
+    zebra_scanner.expect_stdout_line_matches(
+        r"Scanning the blockchain for key 0, started at block 419200, now at block 420000",
+    )?;
+
+    // Make sure scanner scans a few blocks.
+    zebra_scanner.expect_stdout_line_matches(
+        r"Scanning the blockchain for key 0, started at block 419200, now at block 430000",
+    )?;
+    zebra_scanner.expect_stdout_line_matches(
+        r"Scanning the blockchain for key 0, started at block 419200, now at block 440000",
+    )?;
+
+    // Kill the scanner.
+    zebra_scanner.kill(true)?;
+
+    // Check that scan task started and the first scanning is done.
+    let output = zebra_scanner.wait_with_output()?;
+
+    // Make sure the command was killed
+    output.assert_was_killed()?;
+    output.assert_failure()?;
+
+    // Start the node again with the same arguments.
+    let mut zebra_scanner: TestChild<TempDir> = testdir()?.spawn_child(args, false)?;
+
+    // Resuming message.
+    zebra_scanner.expect_stdout_line_matches(
+        "Last scanned height for key number 0 is 439000, resuming at 439001",
+    )?;
+    zebra_scanner.expect_stdout_line_matches("loaded Zebra scanner cache")?;
+
+    // Start scanning where it was left.
+    zebra_scanner.expect_stdout_line_matches(
+        r"Scanning the blockchain for key 0, started at block 439001, now at block 440000",
+    )?;
+    zebra_scanner.expect_stdout_line_matches(
+        r"Scanning the blockchain for key 0, started at block 439001, now at block 450000",
+    )?;
+
+    // Kill the scanner.
+    zebra_scanner.kill(true)?;
+
+    // Check that scan task started and the second scanning is done.
+    let output = zebra_scanner.wait_with_output()?;
+
+    // Make sure the command was killed
+    output.assert_was_killed()?;
+    output.assert_failure()?;
+
+    Ok(())
+}
+
+/// Create a temporary directory for testing `zebra-scanner`.
+pub fn testdir() -> eyre::Result<TempDir> {
+    tempfile::Builder::new()
+        .prefix("zebra_scanner_tests")
+        .tempdir()
+        .map_err(Into::into)
+}
+
+/// Extension trait for methods on `tempfile::TempDir` for using it as a test
+/// directory for `zebra-scanner`.
+pub trait ZebradTestDirExt
+where
+    Self: AsRef<Path> + Sized,
+{
+    // Zebra-scanner methods
+
+    /// Spawn `zebra-scanner` (`zebrad` flag = `false`) or `zebrad` (`zebrad` flag = `true`)
+    /// with `args` as a child process in this test directory.
+    fn spawn_child(self, args: Arguments, zebrad: bool) -> Result<TestChild<Self>>;
```

Comment:
Optional: Could we rename this trait and add `spawn_zebrad_child()`/`spawn_scanner_child()` methods?

```suggestion
pub trait ScannerTestDirExt
where
    Self: AsRef<Path> + Sized,
{
    // Zebra-scanner methods

    /// Spawn `zebra-scanner` with `args` as a child process in this test directory.
    fn spawn_scanner_child(self, args: Arguments) -> Result<TestChild<Self>>;
```

Update: Renamed in https://github.com/ZcashFoundation/zebra/pull/8659/commits/d70d6735d401229e75b24b9c06b971d668b7f5ba.